homeassistant/components/growatt/sensor.py

```diff
@@ -0,0 +1,179 @@
+        )
+        sensor_total = GrowattPlantTotals(
+            hass,
+            growatt_client,
+            username,
+            password,
+            "Growatt plant total",
+            "totalEnergySum",
+        )
+        sensor_current = GrowattPlantCurrent(
+            hass, growatt_client, username, password
+        )
+        add_entities([sensor_today, sensor_total, sensor_current])
+
+
+def login(client, username, password):
+    """Login to the growatt server."""
+    import growatt
+
+    try:
+        client.login(username, password)
+        return True
+    except growatt.LoginError as error:
+        _LOGGER.error(error)
+        return False
+
+
+class GrowattPlant(Entity):
+    """Base Growatt sensor class."""
+
+    def __init__(self, hass, client, username, password):
+        """Initialize the sensor."""
+        self._hass = hass
+        self._client = client
+        self._username = username
+        self._password = password
+        self._state = None
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @staticmethod
+    def convert_multiplier(metric_name, multiplier_lookup):
+        """Convert a value to a given multiplier."""
+        if metric_name not in multiplier_lookup:
+            message = (
+                "Found an unsupported metric name {}"
+                "cannot convert safely to kWh."
+            ).format(metric_name)
+            _LOGGER.error(message)
+            raise ValueError(message)
+        return multiplier_lookup[metric_name]
+
+
+class GrowattPlantTotals(GrowattPlant):
+    """Representation of a Growatt plant sensor."""
+
+    def __init__(self, hass, client, username, password, name, metric_name):
+        """Initialize the sensor."""
+        super().__init__(hass, client, username, password)
+        self._name = name
+        self._metric_name = metric_name
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit this state is expressed in."""
+        return ENERGY_KILO_WATT_HOUR
+
+    def _convert_to_kwh(self, value: str, metric_name: str):
+        """Convert a value to a kWh value."""
+        watts = float(value)
+        multiplier_lookup = {
+            ENERGY_KILO_WATT_HOUR: 1,
+            ENERGY_MEGA_WATT_HOUR: 1000,
+            ENERGY_GIGA_WATT_HOUR: 1000 * 1000,
+        }
+        return watts * self.convert_multiplier(metric_name, multiplier_lookup)
+
+    def _get_total_energy(self, key: str):
+        """Get todays energy as float in kWh.
+
+        Refreshes login to update the session.
+        """
+        login(self._client, self._username, self._password)
+        plant_info = self._client.plant_list()
+        return float(
+            self._convert_to_kwh(*plant_info["totalData"][key].split(" "))
+        )
+
+    def update(self):
+        """Get the latest data from Growatt server."""
+        self._state = self._get_total_energy(self._metric_name)
```

Comment:
As mentioned above, don't raise exceptions in `update` by design.  The current code would allow a `ValueError` to propagate up.  Instead, catch the `HomeAssistantError` here and log it as an error.